cpp/src/decisiontree/batched-levelalgo/kernels/builder_kernels.cuh

```diff
@@ -111,6 +122,209 @@ HDI IdxT lower_bound(DataT* array, IdxT len, DataT element)
+{
+  if (blockIdx.x >= work_items_size) return;
+
+  const uint32_t nodeid = work_items[blockIdx.x].idx;
+
+  uint64_t subsequence(fnv1a32_basis);
+  subsequence = fnv1a32(subsequence, uint32_t(threadIdx.x));
+  subsequence = fnv1a32(subsequence, uint32_t(treeid));
+  subsequence = fnv1a32(subsequence, uint32_t(nodeid));
+
+  raft::random::PCGenerator gen(seed, subsequence, uint64_t(0));
+  raft::random::UniformIntDistParams<IdxT, uint64_t> uniform_int_dist_params;
+
+  uniform_int_dist_params.start = 0;
+  uniform_int_dist_params.end   = n;
+  uniform_int_dist_params.diff =
+    uint64_t(uniform_int_dist_params.end - uniform_int_dist_params.start);
+
+  IdxT n_uniques = 0;
+  IdxT items[MAX_SAMPLES_PER_THREAD];
+  IdxT col_indices[MAX_SAMPLES_PER_THREAD];
+  IdxT mask[MAX_SAMPLES_PER_THREAD];
+  // populate this
+  for (int i = 0; i < MAX_SAMPLES_PER_THREAD; ++i)
+    mask[i] = 0;
+
+  do {
+    // blocked arrangement
+    for (int cta_sample_idx = MAX_SAMPLES_PER_THREAD * threadIdx.x, thread_local_sample_idx = 0;
+         thread_local_sample_idx < MAX_SAMPLES_PER_THREAD;
+         ++cta_sample_idx, ++thread_local_sample_idx) {
+      // mask of the previous iteration, if exists, is re-used here
+      // so previously generated unique random numbers are used.
+      // newly generated random numbers may or may not duplicate the previously generated ones
+      // but this ensures some forward progress in order to generate atleast 'k' unique random
+      // samples.
+      if (mask[thread_local_sample_idx] == 0 and cta_sample_idx < n_parallel_samples)
+        raft::random::custom_next(
+          gen, &items[thread_local_sample_idx], uniform_int_dist_params, IdxT(0), IdxT(0));
+      else if (mask[thread_local_sample_idx] ==
+               0)  // indices that exceed `n_parallel_samples` will not generate
+        items[thread_local_sample_idx] = n - 1;
+      else
+        continue;  // this case is for samples whose mask == 1 (saving previous iteraion's random
+                   // number generated)
+    }
+
+    // Specialize BlockRadixSort type for our thread block
+    typedef cub::BlockRadixSort<IdxT, BLOCK_THREADS, MAX_SAMPLES_PER_THREAD> BlockRadixSortT;
+    // BlockAdjacentDifference
+    typedef cub::BlockAdjacentDifference<IdxT, BLOCK_THREADS> BlockAdjacentDifferenceT;
+    // BlockScan
+    typedef cub::BlockScan<IdxT, BLOCK_THREADS> BlockScanT;
+
+    // Shared memory
+    __shared__ union TempStorage {
+      typename BlockRadixSortT::TempStorage sort;
+      typename BlockAdjacentDifferenceT::TempStorage diff;
+      typename BlockScanT::TempStorage scan;
+    } temp_storage;
+
+    // collectively sort items
+    BlockRadixSortT(temp_storage.sort).Sort(items);
+
+    __syncthreads();
+
+    // compute the mask
+    // compute the adjacent differences according to the functor
+    BlockAdjacentDifferenceT(temp_storage.diff)
+      .FlagHeads(mask, items, mask, CustomDifference<IdxT>());
```

Comment:
We have to wait for `SubtractLeft` to be available in the `cub` that is included in the toolkit. For now, I have added a TODO comment for the `FlagHeads` replacement.
@tfeher Let me know if you think otherwise.